---
title: 'scRNA-seq data analysis'
date: 2021-05-16
categories: [Projects, Genetics]
description: Workflow of scRNA-seq data analysis using R toolkit Seurat. 
tags: [Bioinformatics,scRNA,Transcriptomics,R,Seurat]
pin: false
math: true
---

Single-cell RNA sequencing (scRNA-Seq) is a powerful tool for uncovering cellular heterogeneity within organs and tissues, which is difficult to achieve using bulk RNA sequencing. A widely used analysis toolkit for scRNA-Seq is [Seurat](https://satijalab.org/seurat/) in R. Below, I outline the workflow used in this study.

In this article, I analyzed scRNA-Seq data from 1,278 mouse pancreatic cells and identified marker genes specific to each exocrine and endocrine cell cluster. These included rare pancreatic polypeptide (PP) cells, pancreatic stellate (PS) cells, endothelial cells, and MHC class II antigen-presenting cells. Additionally, the analysis revealed subpopulations of $$\alpha$$, $$\beta$$, $$\delta$$, acinar, and ductal cells. This workflow can be adapted to other datasets for studying cellular heterogeneity.

> The pancreas plays a crucial role in both the digestive and endocrine systems. The exocrine pancreas primarily consists of acinar and ductal cells, responsible for enzyme synthesis and transport, respectively. The endocrine pancreas, known as the islets of Langerhans, is composed mainly of $$\alpha$$, $$\beta$$, PP, $$\delta$$, and $$\epsilon$$ cells, which secrete glucagon, insulin, pancreatic polypeptide (PP), somatostatin, and ghrelin, respectively. Notably, cellular heterogeneity within the pancreas, especially within the islets of Langerhans, has significant functional implications.
{: .prompt-info }


### Import Some Useful Packages.


```r
library(Seurat)
library(patchwork)
library(Matrix)
library(hdf5r)
library(stringr)
library(dplyr)
```

### Import Datasets

```r
setwd() #Change to your working directory
data<-read.csv("Pancreas-counts.csv",row.names= 1)
annotation<-read.csv("pancreas_annotations_FACS.csv")
metadata<-read.csv("pancreas_FACS_metadata.csv")
```

The annotation chart may need to be reshaped, please refer to the documentation of Seurat.

### Pretreatment
#### Filter ERCC

```r
erccs = grep('^ERCC-', x= rownames(x=data),value = T)
percent.ercc = Matrix::colSums(data[erccs, ])/Matrix::colSums(data)
ercc.index = grep(pattern = "^ERCC-", x = rownames(x = data), value = FALSE)
data = data[-ercc.index,]
```

For mitochondrial and ribosomal genomes, the operations are similar.

#### Create Seurat Objects

```r
data = CreateSeuratObject(counts = data, project = "data", min.cells = 3, min.features = 200)
```

#### Quality Control

```r
VlnPlot(data, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
summary(data[["nFeature_RNA"]])
summary(data[["nCount_RNA"]])
summary(data[["percent.mt"]])
plot1 = FeatureScatter(data, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 = FeatureScatter(data, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
```

There was no mitochondrial genome in my data set, so the parameters of the violin plot were somewhat different.

![Fig1](/assets/img/1626072283924.png){: width="470" height="476" }

Seems nFeature_RNA is enriched in [200, 7500].

```r
data = subset(data, subset = nFeature_RNA > 200 & nFeature_RNA <7500 & percent.mt < 10)
dim(data)
```


### Normalization

```r
data = NormalizeData(data, normalization.method = "LogNormalize", scale.factor = 10000)
```

### Feature Selection

```r
data = FindVariableFeatures(data, selection.method = "vst", nfeatures = 2000)
top10 = head(VariableFeatures(data), 10)
plot1 = VariableFeaturePlot(data)
plot2 = LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
```

![Fig2](/assets/img/1626072530925.png){: width="470" height="476" }

### Scale the Data

```r
all.genes = rownames(data)
data = ScaleData(data, features = all.genes)
```

### Perform Linear Dimensional Reduction

```r
data = RunPCA(data, features = VariableFeatures(object = data))
print(data[["pca"]], dims = 1:5, nfeatures = 5)
VizDimLoadings(data, dims = 1:2, reduction = "pca")
DimPlot(data, reduction = "pca")
```

![Fig3](/assets/img/1626072612484.png){: width="470" height="476" }

![Fig4](/assets/img/1626072620017.png){: width="470" height="476" }


### Determine the PC

```r
ElbowPlot(data)
```

![Fig5](/assets/img/1626072656602.png){: width="470" height="476" }

11 clusters seems to be good.

### Cluster the Cells

```r
data = FindNeighbors(data, dims = 1:11)
data = FindClusters(data)
head(Idents(data), 5)
```

It returns me 14 clusters.

### Run Non-linear Dimensional Reduction (UMAP)

```r
data = RunUMAP(data, dims = 1:14)
DimPlot(data, reduction = "umap", label = T)
```

![Fig6](/assets/img/1626072785581.png){: width="470" height="476" }

### Find the Markers

```r
markers <- FindAllMarkers(data, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25,test.use = "roc")
markers=markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
DT::datatable(markers)
```

I think the dotplot below is useful if you are sure of the cell types that may be present in the sample. I have listed here the marker genes of pancreatic cells (both exocrine and endocrine).

```r
DotPlot(data, features = c("Gcg","Ins1","Ins2","Iapp","Ppy","Sst","Ghrl","Prss1","Krt19","Plvap","Pecam1","Vwf","Angpt2","Col1a2","Timp1","Fn1","Postn","Acta2","Tpsb2","Tpsd1","Tpsab1","Cd86"))
```

![Fig7](/assets/img/1626073344686.png){: width="1200" height="476" }

### Rename the Clusters

```r
cluster.ids <- c("Beta cells","Co-expression","Acinar cells","Beta cells","Ductal cells","Delta cells","Acinar cells","Co-expression","Unclassified cells","Unclassified cells","Co-expression","Endothelial cells","MHC class 2","PSCs")
names(cluster.ids) <- levels(data)
data <- RenameIdents(data, cluster.ids)
DimPlot(data, reduction = "umap", label = TRUE, pt.size = 1)
```

(Updated on 2021/11/16) There is a problem with the axis of original figure. Now that the data from exocrine and endocrine are separated... the results of UMAP are as follows.

#### Exocrine

![Fig8](/assets/img/1637048141277.png){: width="470" height="470" }

#### Endocrine

![Fig9](/assets/img/1637048147444.png){: width="470" height="470" }

### What's the next step?

The differences in physiological functions between the marker genes of different groups can be examined using online Gene Ontology (GO) enrichment analysis. 

